swift/Sources/FlatBuffers/Verifier.swift

```diff
@@ -0,0 +1,181 @@
+      return
+    }
+    throw FlatbuffersErrors.missAlignedPointer(
+      position: position,
+      type: String(describing: T.self))
+  }
+
+  ///
+  /// - Parameters:
+  ///   - position:
+  ///   - size:
+  /// - Returns:
+  public mutating func rangeInBuffer(position: Int, size: Int) throws {
+    let end = position &+ size
+    #warning("Check if this is correct")
+    if end < 0 || end > _buffer.capacity {
+      throw FlatbuffersErrors.outOfBounds(position: position, end: end)
+    }
+    _apparentSize = _apparentSize &+ UInt32(size)
+    if _apparentSize > _options._maxApparentSize {
+      throw FlatbuffersErrors.apparentSizeTooLarge
+    }
+  }
+
+  ///
+  /// - Parameters:
+  ///   - position:
+  ///   - type:
+  /// - Returns:
+  public mutating func inBuffer<T>(position: Int, of type: T.Type) throws {
+    try isAligned(position: position, type: type)
+    try rangeInBuffer(position: position, size: MemoryLayout<T>.size)
+  }
+
+  ///
+  /// - Parameter position:
+  /// - Returns:
+  public mutating func visitTable(at position: Int) throws -> TableVerifier {
+    let vtablePosition = try derefOffset(position: position)
+    let vtableLength: VOffset = try getValue(at: vtablePosition)
+
+    let length = Int(vtableLength)
+    try isAligned(position: vtablePosition + length, type: VOffset.self)
+    try rangeInBuffer(position: vtablePosition, size: length)
+
+    _tableCount += 1
+
+    if _tableCount > _options._maxTableCount {
+      throw FlatbuffersErrors.maximumTables
+    }
+
+    _depth += 1
+
+    if _depth > _options._maxDepth {
+      throw FlatbuffersErrors.maximumDepth
+    }
+
+    return TableVerifier(
+      position: position,
+      vtable: vtablePosition,
+      vtableLength: length,
+      verifier: &self)
+  }
+
+  ///
+  /// - Parameter position:
+  /// - Returns:
+  internal mutating func getValue<T>(at position: Int) throws -> T {
+    try inBuffer(position: position, of: T.self)
+    return _buffer.read(def: T.self, position: position)
+  }
+
+  internal mutating func derefOffset(position: Int) throws -> Int {
+    try inBuffer(position: position, of: Int32.self)
+
+    let offset = _buffer.read(def: Int32.self, position: position)
+    // switching to int32 since swift's default Int is int64
+    let _int32Position = Int32(position)
+
+    let reportedOverflow: (partialValue: Int32, overflow: Bool)
+    if offset > 0 {
+      reportedOverflow = _int32Position.subtractingReportingOverflow(abs(offset))
+    } else {
+      reportedOverflow = _int32Position.addingReportingOverflow(abs(offset))
+    }
+
+    /// since `subtractingReportingOverflow` & `addingReportingOverflow` returns true,
+    /// if there is overflow we return failure
+    #warning("see if we should always be positive here")
+    if reportedOverflow.overflow || reportedOverflow.partialValue > _buffer.capacity || reportedOverflow.partialValue <= 0 {
+      throw FlatbuffersErrors.signedOffsetOutOfBounds(
```

Comment:
I've kinda copied the rust code in here, however it seems to produce a negative value example `-2` is this allowed? or does it have to strictly be positive number?

This is done like this since I tried to fuzz the verifier, and it seemed to crash there each time

Reply:
Hmm this looks very different from the C++ code

Reply:
As i said earlier this code is similar to that of rust. since it was the latest to be added and most similar to swift

Reply:
@aardappel 
After looking into the code and checking the rust code again, and running the fuzzer again. At the begining i thought that it was an overflow but it doesnt seem to be the case since `reportedOverflow.overflow` is `false` and the fuzzer is throwing an `addressSanitizer: SEGV on unknown address` error since we are trying to read a negative value `-2`, `-1073741820` which is clearly out of the bounds of the buffer.
maybe its also an issue in the rust code @CasperN?

Reply:
So the solution was interesting, apparently `abs()` in swift can't handle already overflowing values thus the issue that we faced. and the solution was basically to get the `magnitude` of the value so

```
var int32: Int32 = -2147483648

print(int32.magnitude)
// prints 2147483648

print(abs(int32))
// prints -2147483648
```